MarkLogic New Integration (#6524)
DataDog/integrations-core: marklogic/README.md

# Agent Check: MarkLogic

## Overview

This check monitors [MarkLogic][1] through the Datadog Agent.

## Setup

Follow the instructions below to install and configure this check for an Agent running on a host. For containerized environments, see the [Autodiscovery Integration Templates][2] for guidance on applying these instructions.

### Installation

The MarkLogic check is included in the [Datadog Agent][2] package.
No additional installation is needed on your server.

#### Prepare MarkLogic

> Using the API or the Admin interface, create a user for the Datadog Agent with at least the [`manage-admin`][3] role.

Reviewer: `manage-user` role is not enough if we want to access the `/manage/v2?view=health` endpoint.

Reviewer: That's expected to work with `datadog:datadog` because it's already a `manage-admin` (see https://github.com/DataDog/integrations-core/pull/6524/files#diff-eb070e813473dda6f8d12fc351573c38R67)

If you spin up the e2e environment create a `manage-user` and try to access this endpoint, you should get an error.
```
curl -X POST --anyauth --user admin:admin -i -H "Content-Type: application/json" -d '{"user-name": "datadog2", "password": "datadog", "roles": {"role": "manage-user"}}' http://localhost:8002/manage/v2/users
curl -X GET --anyauth --user datadog2:datadog -i http://localhost:8002/manage/v2\?view\=health\&format\=json
```

Reviewer: The health service checks are now behind a configuration option `enable_health_service_checks` disable by default.
When enabled, it needs the `manage-admin` permission, otherwise `manage-user` is enough.

# Agent Check: MarkLogic

## Overview

> This check monitors [MarkLogic][1] through the Datadog Agent.

Reviewer 2: ```suggestion
This check monitors [MarkLogic][1] through the Datadog Agent. MarkLogic Server is a multi-model database designed to be a data hub for operational and analytical data.
```

Include a line about MarkLogic, maybe even overview of the types of metrics collected?

## Setup

Follow the instructions below to install and configure this check for an Agent running on a host. For containerized environments, see the [Autodiscovery Integration Templates][2] for guidance on applying these instructions.

### Installation

The MarkLogic check is included in the [Datadog Agent][2] package.
No additional installation is needed on your server.

#### Prepare MarkLogic

> Using the API or the Admin interface, create a user for the Datadog Agent with at least the [`manage-user`][3] role.

Reviewer 2: ```suggestion
Using the API or the Admin interface, create a user for the Datadog Agent with the [`manage-user`][3] role permissions at minimum.
```

> In order to use the `enable_health_service_checks` configuration, the Datadog MarkLogic user should have at least the [`manage-admin`][4] role

Reviewer 2: ```suggestion
In order to use the `enable_health_service_checks` configuration, the Datadog MarkLogic user should have at least the [`manage-admin`][4] role.
```

##### Using the API

1. Create the Datadog user with a request like:
    ```shell
    curl -X POST --anyauth --user <ADMIN_USER>:<ADMIN_PASSWORD> -i -H "Content-Type: application/json" -d '{"user-name": "<USER>", "password": "<PASSWORD>", "roles": {"role": "manage-user"}}' http://<HOSTNAME>:8002/manage/v2/users
    ```
    Use the correct `<ADMIN_USER>` and `<ADMIN_PASSWORD>`, and replace `<USER>` and `<PASSWORD>` with what the Datadog Agent will use.

>     Full documentation about the endpoint [here][5].

Reviewer 2: ```suggestion
    For more information on the endpoint, see [documentation][5].
```

2. Confirm the user was created with enough permissions with a request like:
    ```shell
    curl -X GET --anyauth --user <USER>:<PASSWORD> -i http://<HOSTNAME>:8002/manage/v2
    ```

##### Using the Admin interface

> 1. Go to the QConsole with an admin account. By default it's at http://<HOSTNAME>:8000/qconsole.

Reviewer 2: ```suggestion
1. Go to the QConsole with an admin account. By default, the QConsole is available at http://<HOSTNAME>:8000/qconsole.
```

2. Select `Security` as Database and `XQuery` as query type.

3. Run this query, replacing `<USER>` and `<PASSWORD>` with what the Datadog Agent will use:
    ```
    xquery version "1.0-ml";
    import module namespace sec="http://marklogic.com/xdmp/security" at 
        "/MarkLogic/security.xqy";

    sec:create-user(
        "<USER>",
        "Datadog Agent user",
        "<PASSWORD>",
        "manage-user",
        (xdmp:permission("security", "read")),
        ("http://marklogic.com/dev_modules"))
    
    ```

>     Full documentation about the query [here][6]

Reviewer 2: ```suggestion
   For more information on the query, see [documentation][6].
```

4. Confirm the user was created with enough permissions using `<USER>` and `<PASSWORD>` to authenticate at http://<HOSTNAME>:8002 (default port).

### Configuration

> 1. Edit the `marklogic.d/conf.yaml` file, in the `conf.d/` folder at the root of your Agent's configuration directory to start collecting your marklogic performance data. See the [sample marklogic.d/conf.yaml][7] for all available configuration options. Use the Datadog Agent user previously created in the configuration file.

Reviewer 2: ```suggestion
1. Edit the `marklogic.d/conf.yaml` file, in the `conf.d/` folder at the root of your Agent's configuration directory to start collecting your MarkLogic performance data. See the [sample marklogic.d/conf.yaml][7] for all available configuration options. Use the Datadog Agent user previously created in the configuration file.
```

2. [Restart the Agent][8].

### Validation

[Run the Agent's status subcommand][9] and look for `marklogic` under the Checks section.

## Data Collected

### Metrics

See [metadata.csv][10] for a list of metrics provided by this check.

### Service Checks

> MarkLogic does not include any service checks.

Reviewer 2: Include `marklogic.database.health`, `marklogic.forest.health`, and `marklogic.can_connect`.
